asap2wav.1
.TH ASAP2WAV 1 "December 8, 2008"
.SH NAME
asap2wav \- convert 8-bit Atari music to WAV files
.SH SYNOPSIS
.B asap2wav
.B [\-o
.I FILE
.B ] [\-s
.I SONG
.B ] [\-t
.I TIME
.B ] [\-r
.I FREQ
.B ] [\-q
.I QUALITY
.B ] [\-b] [\-w] [\-\-raw]
.I INPUTFILE...
.SH DESCRIPTION
.B asap2wav
converts 8-bit Atari music files to WAV files. It internally
emulates the POKEY sound chip and the 6502 processor.
.SH ARGUMENTS
Format of the input files is detected from their filename extensions.
Currently the following formats are supported:
.TP
SAP
Slight Atari Player - standard file format for playing 8-bit Atari music
on modern computers.
.TP
CMC
Chaos Music Composer.
.TP
CMR
CMC Rzog - CMC with modified bass sounds.
.TP
DMC
DoublePlay CMC - CMC with 6502 routines executed at double rate
of the original CMC.
.TP
MPT
Music ProTracker. Samples are not supported yet.
.TP
MPD
MPT DoublePlay - MPT with 6502 routines executed at double rate
of the original MPT.
.TP
RMT
Raster Music Tracker.
.TP
TMC, TM8
Theta Music Composer 1.x
.TP
TM2
Theta Music Composer 2.x.
.SH OPTIONS
.TP
\fB\-o\fR \fIFILE\fR, \fB\-\-output\fR=\fIFILE\fR
Set output WAV file name. "-" means standard output.
If you do not specify this option, the name of the input file is reused,
with the extension changed to "wav".
.TP
\fB\-s\fR \fISONG\fR, \fB\-\-song\fR=\fISONG\fR
Select subsong number (zero-based).
.TP
\fB\-t\fR \fITIME\fR, \fB\-\-time\fR=\fITIME\fR
Set output length in the "MM:SS" format.
.TP
\fB\-b\fR, \fB\-\-byte-samples\fR
Output 8-bit samples.
.TP
\fB\-w\fR, \fB\-\-word-samples\fR
Output 16-bit samples (default).
.TP
\fB\-\-raw\fR
Output audio samples without the WAV header.
.TP
\fB\-m\fR \fICHANNELS\fR, \fB\-\-mute\fR=\fICHANNELS\fR
Mute POKEY channels specified by digits: 1-4 for the base (mono or left) POKEY,
5-8 for the extra (right) POKEY.
.TP
\fB\-h\fR, \fB\-\-help\fR
Display help message and exit.
.TP
\fB\-v\fR, \fB\-\-version\fR
Display version information and exit.
.SH COPYRIGHT
ASAP2WAV is free software; you can redistribute it and/or modify it
under the terms of the GNU General Public License as published
by the Free Software Foundation; either version 2 of the License,
or (at your option) any later version.


NEWS
History
-------

* ASAP 1.2.0 (2008-03-29)

Plugins for Windows Media Player and Apollo have been created.
RMT duration detection has been improved. A bug in TMC to SAP conversion
has been fixed. ASAP can be compiled as C#.


* ASAP 1.1.0 (2008-02-17)

All formats supported by ASAP can be converted to and from the SAP format,
using the file information window of WASAP or the Winamp plugin.
Unlike SAP Maker, ASAP supports subsongs and TIME tags.
POKEY emulation has improved a little. Bugs in the XMMS plugin, the foobar2000
plugin and the applet have been fixed.


* ASAP 1.0.2 (2007-11-02)

TIME tags can be edited in the file information window of WASAP
and the Winamp plugin. A few emulation bugs have been fixed.
Some user interface issues in WASAP have been fixed.


* ASAP 1.0.1 (2007-06-10)

Plugin for MOC (Music On Console) has been created. WASAP, Winamp plugin
and foobar2000 plugin can edit AUTHOR, NAME and DATE tags in SAP files.
Winamp and XMMS plugins format titles as specified by the player.
TMC files are now played in stereo. A few minor bugs have been fixed.


* ASAP 1.0.0 (2007-05-28)

This release introduces completely new 6502 and POKEY emulation.
SAP files with samples are supported. ASAP can be compiled as Java - an applet
and an experimental midlet have been created.
Winamp, foobar2000 and GSPlayer plugins support silence detection
and muting selected POKEY channels. Winamp plugin supports subsongs
and Winamp's equalizer.


* ASAP 0.3.0 (2007-05-06)

This release adds support for AUTHOR, NAME, DATE and TIME tags in SAP files.
Winamp, foobar2000 and the new GSPlayer plugin have configurable playback time
for songs with no TIME tag and songs marked with LOOP.
foobar2000 plugin has been updated for version 0.9 of foobar2000.
XMMS plugin supports the TM8 extension.


* ASAP 0.2.1 (2006-01-28)

This release supports TM2 (Theta Music Composer 2.x) file format.
Compilation and installation on Unix-like systems have been improved.
foobar2000 plugin supports subsongs.


* ASAP 0.2.0 (2006-01-15)

This release supports stereo and 16-bit audio output (enabled by default).
Compressed SAP files (for example Fractals.sap) are played correctly.
Subsongs are supported in RMT files.
There are new plugins: for XMMS and for foobar2000 version 0.8.
WASAP supports Windows Explorer's "Open with".


* ASAP 0.1.0 (2005-12-31)

This is the initial release of ASAP (Another Slight Atari Player)
- a player/converter of 8-bit Atari music for modern computers.
ASAP provides high-quality, identical to the popular Atari800 emulator,
emulation of the POKEY sound chip and the 6502 processor.

This release includes: ASAP2WAV - portable command-line utility for converting
8-bit Atari music to WAV files, WASAP - a tiny player for Windows
and in_asap - a Winamp plugin. Currently supported input formats are:
SAP (except for samples and stereo), CMC, CMR, DMC, MPT (without samples),
MPD, RMT (mono) and TMC (mono).


asap2wav.c
/*
 * asap2wav.c - converter of ASAP-supported formats to WAV files
 *
 * Copyright (C) 2005-2008  Piotr Fusik
 *
 * This file is part of ASAP (Another Slight Atari Player),
 * see http://asap.sourceforge.net
 *
 * ASAP is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published
 * by the Free Software Foundation; either version 2 of the License,
 * or (at your option) any later version.
 *
 * ASAP is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty
 * of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with ASAP; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdarg.h>

#include "asap.h"

static abool no_input_files = TRUE;
static const char *output_file = NULL;
static abool output_header = TRUE;
static int song = -1;
static abool use_16bit = TRUE;
static int duration = -1;
static int mute_mask = 0;

static void print_help(void)
{
	printf(
		"Usage: asap2wav [OPTIONS] INPUTFILE...\n"
		"Each INPUTFILE must be in a supported format:\n"
		"SAP, CMC, CMR, DMC, MPT, MPD, RMT, TMC, TM8 or TM2.\n"
		"Options:\n"
		"-o FILE     --output=FILE      Set output file name\n"
		"-o -        --output=-         Write to standard output\n"
		"-s SONG     --song=SONG        Select subsong number (zero-based)\n"
		"-t TIME     --time=TIME        Set output length (MM:SS)\n"
		"-b          --byte-samples     Output 8-bit samples\n"
		"-w          --word-samples     Output 16-bit samples (default)\n"
		"            --raw              Output raw audio (no WAV header)\n"
		"-m CHANNELS --mute=CHANNELS    Mute POKEY channels (1-8)\n"
		"-h          --help             Display this information\n"
		"-v          --version          Display version information\n"
	);
	no_input_files = FALSE;
}

static void print_version(void)
{
	printf("ASAP2WAV " ASAP_VERSION "\n");
	no_input_files = FALSE;
}

static void fatal_error(const char *format, ...)
{
	va_list args;
	va_start(args, format);
	fprintf(stderr, "asap2wav: ");
	vfprintf(stderr, format, args);
	fputc('\n', stderr);
	va_end(args);
	exit(1);
}

static void set_song(const char *s)
{
	song = 0;
	do {
		if (*s < '0' || *s > '9')
			fatal_error("subsong number must be an integer");
		song = 10 * song + *s++ - '0';
		if (song > 31)
			fatal_error("maximum subsong number is 31");
	} while (*s != '\0');
}

static void set_time(const char *s)
{
	duration = ASAP_ParseDuration(s);
	if (duration <= 0)
		fatal_error("invalid time format");
}

static void set_mute_mask(const char *s)
{
	int mask = 0;
	while (*s != '\0') {
		if (*s >= '1' && *s <= '8')
			mask |= 1 << (*s - '1');
		s++;
	}
	mute_mask = mask;
}

/* write 16-bit word as little endian */
static void fput16(int x, FILE *fp)
{
	fputc(x & 0xff, fp);
	fputc((x >> 8) & 0xff, fp);
}

/* write 32-bit word as little endian */
static void fput32(int x, FILE *fp)
{
	fputc(x & 0xff, fp);
	fputc((x >> 8) & 0xff, fp);
	fputc((x >> 16) & 0xff, fp);
	fputc((x >> 24) & 0xff, fp);
}

static void process_file(const char *input_file)
{
	FILE *fp;
	static byte module[ASAP_MODULE_MAX];
	int module_len;
	static ASAP_State asap;
	int n_bytes;
	static byte buffer[8192];
	if (strlen(input_file) >= FILENAME_MAX)
		fatal_error("filename too long");
	fp = fopen(input_file, "rb");
	if (fp == NULL)
		fatal_error("cannot open %s", input_file);
	module_len = fread(module, 1, sizeof(module), fp);
	fclose(fp);
	if (!ASAP_Load(&asap, input_file, module, module_len))
		fatal_error("%s: format not supported", input_file);
	if (song < 0)
		song = asap.module_info.default_song;
	if (song >= asap.module_info.songs) {
		fatal_error("you have requested subsong %d ...\n"
			"... but %s contains only %d subsongs",
			song, input_file, asap.module_info.songs);
	}
	if (duration < 0) {
		duration = asap.module_info.durations[song];
		if (duration < 0)
			duration = 180 * 1000;
	}
	ASAP_PlaySong(&asap, song, duration);
	ASAP_MutePokeyChannels(&asap, mute_mask);
	if (output_file == NULL) {
		static char output_default[FILENAME_MAX];
		strcpy(output_default, input_file);
		ASAP_ChangeExt(output_default, output_header ? "wav" : "raw");
		output_file = output_default;
	}
	if (output_file[0] == '-' && output_file[1] == '\0')
		fp = stdout;
	else {
		fp = fopen(output_file, "wb");
		if (fp == NULL)
			fatal_error("cannot write %s", output_file);
	}
	if (output_header) {
		int block_size = asap.module_info.channels << use_16bit;
		int bytes_per_second = ASAP_SAMPLE_RATE * block_size;
		n_bytes = duration * (ASAP_SAMPLE_RATE / 100) / 10 * block_size;
		fwrite("RIFF", 1, 4, fp);
		fput32(n_bytes + 36, fp);
		fwrite("WAVEfmt \x10\0\0\0\1\0", 1, 14, fp);
		fput16(asap.module_info.channels, fp);
		fput32(ASAP_SAMPLE_RATE, fp);
		fput32(bytes_per_second, fp);
		fput16(block_size, fp);
		fput16(8 << use_16bit, fp);
		fwrite("data", 1, 4, fp);
		fput32(n_bytes, fp);
	}
	do {
		n_bytes = ASAP_Generate(&asap, buffer, sizeof(buffer),
			use_16bit ? ASAP_FORMAT_S16_LE : ASAP_FORMAT_U8);
		if (fwrite(buffer, 1, n_bytes, fp) != n_bytes) {
			fclose(fp);
			fatal_error("error writing to %s", output_file);
		}
	} while (n_bytes == sizeof(buffer));
	if (!(output_file[0] == '-' && output_file[1] == '\0'))
		fclose(fp);
	output_file = NULL;
	song = -1;
	duration = -1;
	no_input_files = FALSE;
}

int main(int argc, char *argv[])
{
	int i;
	for (i = 1; i < argc; i++) {
		const char *arg = argv[i];
		if (arg[0] != '-')
			process_file(arg);
		else if (arg[1] == 'o' && arg[2] == '\0')
			output_file = argv[++i];
		else if (strncmp(arg, "--output=", 9) == 0)
			output_file = arg + 9;
		else if (arg[1] == 's' && arg[2] == '\0')
			set_song(argv[++i]);
		else if (strncmp(arg, "--song=", 7) == 0)
			set_song(arg + 7);
		else if (arg[1] == 't' && arg[2] == '\0')
			set_time(argv[++i]);
		else if (strncmp(arg, "--time=", 7) == 0)
			set_time(arg + 7);
		else if ((arg[1] == 'b' && arg[2] == '\0')
			|| strcmp(arg, "--byte-samples") == 0)
			use_16bit = FALSE;
		else if ((arg[1] == 'w' && arg[2] == '\0')
			|| strcmp(arg, "--word-samples") == 0)
			use_16bit = TRUE;
		else if (strcmp(arg, "--raw") == 0)
			output_header = FALSE;
		else if (arg[1] == 'm' && arg[2] == '\0')
			set_mute_mask(argv[++i]);
		else if (strncmp(arg, "--mute=", 7) == 0)
			set_mute_mask(arg + 7);
		else if ((arg[1] == 'h' && arg[2] == '\0')
			|| strcmp(arg, "--help") == 0)
			print_help();
		else if ((arg[1] == 'v' && arg[2] == '\0')
			|| strcmp(arg, "--version") == 0)
			print_version();
		else
			fatal_error("unknown option: %s", arg);
	}
	if (no_input_files) {
		print_help();
		return 1;
	}
	return 0;
}
